src/kafka/kafka_clients.clj

```diff
@@ -0,0 +1,170 @@
+(ns kafka.kafka-clients
+  (:import (java.net InetAddress)
+           (org.apache.kafka.clients.producer KafkaProducer ProducerRecord)
+           (org.apache.kafka.common.serialization StringSerializer StringDeserializer)
+           (org.apache.kafka.clients.consumer KafkaConsumer)
+           (java.time Duration)))
+
+(defn build-producer-config-atom
+  []
+  (atom {"value.serializer"  StringSerializer
+         "key.serializer"    StringSerializer
+         "client.id"         (.getHostName (InetAddress/getLocalHost))
+         "bootstrap.servers" "localhost:9092"
+         "acks"              "all"}))
+
+(defn build-consumer-config-atom
+  []
+  (atom {"group.id"           "My-Group"
+         "key.deserializer"   StringDeserializer
+         "value.deserializer" StringDeserializer
+         "client.id"          (.getHostName (InetAddress/getLocalHost))
+         "bootstrap.servers"  "localhost:9092"}))
+
+(defn build-producer-atom
+  [config]
+  (let [config (deref config)]
+    (atom (new KafkaProducer config))))
+
+(defn build-consumer-atom
+  [config]
+  (let [config   (deref config)
+        consumer (new KafkaConsumer config)]
+    (.subscribe consumer ["demo"])
+    (atom consumer)))
+
+(defn send-msg!
+  [producer record]
+  (let [producer (deref producer)]
+    (.send producer record)))
+
+(defn read-msg!
+  [consumer]
+  (let [consumer (deref consumer)
+        record-s (.poll consumer (Duration/ofMillis 100))]
+    (.commitAsync consumer)
+    (mapv (fn [record]
+            (str "Processed Value: " (.value record)))
+          record-s)))
+
+(defn build-producer
+  [config]
+  (let [config (deref config)]
+    (new KafkaProducer config)))
+
+(defn build-consumer
+  [config]
+  (let [config   (deref config)
+        consumer (new KafkaConsumer config)]
+    (.subscribe consumer ["demo"])
+    consumer))
```

Comment:
No code outside of the let block!!!!

Reply:
I want to let the consumer subscribe topic and return the consumer, how could I move them to the let binding? The subscribe only return nil, should I use `(let [subscribe-result (.subscribe consumer ["demo"])])` ?

Reply:
See previous comment